documentation/coding_guidelines.md

```diff
@@ -254,7 +254,28 @@ We currently use [log](https://docs.rs/log/) for logging.
 
 *Unit tests*
 
-Ideally, all code should be unit tested.  Unit test files should be in the same directory as `mod.rs` and their file names should end in `_test.rs`.  A module to be tested should have the test modules annotated with `#[cfg(test)]`.  For example, if in a crate there is a db module, the expected directory structure is as follows:
+Ideally, all code should be unit tested.  Unit tests should ideally be in the same file as the code it is testing though in a distinct module, using the following syntax:
+
+```rust
+struct Foo {
+}
+
+impl Foo {
+  pub fn magic_number() -> u8 {
+    42
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  #test
+  fn verify_magic_number() {
+    assert_eq!(Foo::magic_number(), 42);
+  }
+}
+```
+
+It is also acceptable to place test files into the same directory as `mod.rs` or `lib.rs` so long as their file names end in `_test.rs`. `#[cfg(test)]` should guard tests from release builds. For example, if in a crate there is a db module, the expected directory structure is as follows:
```

Comment:
We should remove this paragraph and discourage this type of structure.

Reply:
Fair enough, we don't need the struct, the intent was to show code and tests co-located.

Reply:
I suggest advising to use this pattern if the test module becomes too big.

Reply:
actually @bmwill comment isn't clear, is it on my crappy Foo structure or on line 278. In general, I am not a big fan of line 278. It probably means that your code is too long and complex or that you're not writing a unit test.

Reply:
specifically line 278, test modules should be submodules of the code they are trying to test, keeping the tests close to the source. We should not encourage placing tests is `_test.rs` files at the root of a crate....Though now that I'm reading it again its not exactly saying what I thought it was (which maybe means it wasn't clear enough?) lol

Reply:
I agree with @bmwill that this isn't super clear. Why not just link to the best practices guide instead of trying to define this ourselves? https://doc.rust-lang.org/book/ch11-03-test-organization.html

Or is it because we want to deviate from the guide with exceptional cases?

Reply:
Yes! Let just link to the best practices :)